firmware-binaries/test-cases/scatter_gather_test/src/main.rs

```diff
@@ -0,0 +1,73 @@
+// SPDX-FileCopyrightText: 2024 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+#![no_std]
+#![cfg_attr(not(test), no_main)]
+
+use bittide_sys::gather_unit::GatherUnit;
+use bittide_sys::scatter_unit::ScatterUnit;
+use bittide_sys::uart::Uart;
+use core::fmt::Write;
+#[cfg(not(test))]
+use riscv_rt::entry;
+
+const UART_ADDR: *const () = (0b010 << 29) as *const ();
+const SCATTER_ADDR: *const () = (0b011 << 29) as *const ();
+const GATHER_ADDR: *const () = (0b100 << 29) as *const ();
+
+// Therefore, `MEM_SIZE` is twice as large as the calendar size. `MEM_SIZE` is
+// defined as 32-bit words. For this test a calendar depth of 16 64-bit words is used.
+const MEM_SIZE: usize = 2 * 16;
+
+#[cfg_attr(not(test), entry)]
+fn main() -> ! {
+    // Initialize peripherals.
+    let mut uart = unsafe { Uart::new(UART_ADDR) };
+    let scatter_unit: ScatterUnit<MEM_SIZE> = unsafe { ScatterUnit::new(SCATTER_ADDR) };
+    let gather_unit: GatherUnit<MEM_SIZE> = unsafe { GatherUnit::new(GATHER_ADDR) };
+
+    // For this test an array of incrementing values is written to the full gather
+    // memory. The calendars for the scatter and gather units are expected to write the
+    // data sent over the link to the same memory location in the scatter unit. This makes
+    // this test behave like an 'echo'.
+    // Test takes 3 metacycles:
+    //  1: CPU writes to the gather unit memory
+    //  2: Gather memory is written over the link to the scatter memory
+    //  3: Read from scatter memory
+
+    // First metacycle
+    let source: [u8; MEM_SIZE] = core::array::from_fn(|i| i as u8);
+    gather_unit.wait_for_new_metacycle();
+    unsafe { gather_unit.copy_from_slice(&source, 0) };
+
+    // Second metacycle
+    gather_unit.wait_for_new_metacycle();
+
+    // Third metacycle
+    gather_unit.wait_for_new_metacycle();
+    let mut destination: [u8; MEM_SIZE] = [0; MEM_SIZE];
+    unsafe { scatter_unit.copy_to_slice(destination.as_mut(), 0) };
+
```

Comment:
By implementing `Deref` (with `Target = [u8; MEM_SIZE]`, [link](https://doc.rust-lang.org/std/ops/trait.Deref.html)) and `DerefMut` ([link](https://doc.rust-lang.org/std/ops/trait.DerefMut.html)) for `GatherUnit` and `ScatterUnit`, you can instead use the safe slice method `slice::copy_from_slice` and remove `GatherUnit::copy_from_slice` and `ScatterUnit::copy_to_slice`. Note that line 49 would have to change to `destination.copy_from_slice(&scatter_unit[..])` with this change.

Reply:
Sorry, took someone reminding me that this is MMIO so it needs to be handled a bit differently. I'll write the relevant adjustments in the comment on `copy_from_slice` just below, but this can still have a safe interface.